crates/sui-framework/packages/sui-framework/sources/authenticator_state.move

```diff
@@ -0,0 +1,369 @@
+            return string_bytes_lt(&a.jwk.kty, &b.jwk.kty)
+        };
+        if (&a.jwk.e != &b.jwk.e) {
+            return string_bytes_lt(&a.jwk.e, &b.jwk.e)
+        };
+        if (&a.jwk.n != &b.jwk.n) {
+            return string_bytes_lt(&a.jwk.n, &b.jwk.n)
+        };
+        string_bytes_lt(&a.jwk.alg, &b.jwk.alg)
+    }
+
+    #[allow(unused_function)]
+    /// Create and share the AuthenticatorState object. This function is call exactly once, when
+    /// the authenticator state object is first created.
+    /// Can only be called by genesis or change_epoch transactions.
+    fun create(ctx: &TxContext) {
+        assert!(tx_context::sender(ctx) == @0x0, ENotSystemAddress);
+
+        let version = CurrentVersion;
+
+        let inner = AuthenticatorStateInner {
+            version,
+            active_jwks: vector[],
+        };
+
+        let self = AuthenticatorState {
+            id: object::authenticator_state(),
+            version,
+        };
+
+        dynamic_field::add(&mut self.id, version, inner);
+        transfer::share_object(self);
+    }
+
+    fun load_inner_mut(
+        self: &mut AuthenticatorState,
+    ): &mut AuthenticatorStateInner {
+        let version = self.version;
+
+        // replace this with a lazy update function when we add a new version of the inner object.
+        assert!(version == CurrentVersion, EWrongInnerVersion);
+
+        let inner: &mut AuthenticatorStateInner = dynamic_field::borrow_mut(&mut self.id, self.version);
+
+        assert!(inner.version == version, EWrongInnerVersion);
+        inner
+    }
+
+    fun load_inner(
+        self: &AuthenticatorState,
+    ): &AuthenticatorStateInner {
+        let version = self.version;
+
+        // replace this with a lazy update function when we add a new version of the inner object.
+        assert!(version == CurrentVersion, EWrongInnerVersion);
+
+        let inner: &AuthenticatorStateInner = dynamic_field::borrow(&self.id, self.version);
+
+        assert!(inner.version == version, EWrongInnerVersion);
+        inner
+    }
+
+    fun check_sorted(new_active_jwks: &vector<ActiveJwk>) {
+        let i = 0;
+        while (i < vector::length(new_active_jwks) - 1) {
+            let a = vector::borrow(new_active_jwks, i);
+            let b = vector::borrow(new_active_jwks, i + 1);
+            assert!(jwk_lt(a, b), EJwksNotSorted);
+            i = i + 1;
+        };
+    }
+
+    #[allow(unused_function)]
+    /// Record a new set of active_jwks. Called when executing the AuthenticatorStateUpdate system
+    /// transaction. The new input vector must be sorted and must not contain duplicates.
+    /// If a new JWK is already present, but with a previous epoch, then the epoch is updated to
+    /// indicate that the JWK has been validated in the current epoch and should not be expired.
+    fun update_authenticator_state(
+        self: &mut AuthenticatorState,
+        new_active_jwks: vector<ActiveJwk>,
```

Comment:
How big can this list get?

Reply:
Each validator can submit at most 2400 jwk votes per epoch, and we maintain the current + previous epoch's JWKs, so up to 4800. It is highly unlikely we would have this many. I may reduce the limit in fact, i think 2400 is probably too much.

Reply:
Yes, you probably don't want this getting much more than ~2000 for that total?